File: PlaywireSDKAppsObjC/ads/fullscreen/rewarded/RewardedViewController.m
//
//  RewardedViewController.m
//  PlaywireDemo
//
//  Created by Playwire Mobile Team on 01/05/23.
//  Copyright © 2023 Playwire. All rights reserved.
//

#import "RewardedViewController.h"
#import <Playwire-Swift.h>

@interface RewardedViewController () <PWFullScreenAdDelegate>
@property (weak, nonatomic) IBOutlet UILabel *statusLabel;
@property (weak, nonatomic) IBOutlet UIButton *getRewardButton;
@property (strong, nonatomic) PWRewarded *rewarded;
@end

@implementation RewardedViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.getRewardButton setEnabled:NO];
    [self loadRewarded];
}

- (IBAction)getRewardAction:(UIButton *)sender {
    [self showRewarded];
}

- (void)loadRewarded {
    self.rewarded = [[PWRewarded alloc] initWithAdUnitName:self.adUnitName delegate:self];
    
    // Use `[[PWLoadParams new] withTargeting:]` to pass your custom targets to ad request.
    // PWLoadParams *params = [[PWLoadParams new] withTargeting:@{
    //   @"age": @"18-32",
    //   @"page": @"travel"
    // }];
    // [self.rewarded loadWithParams:params];
    
    [self.rewarded load];
    self.statusLabel.text = [NSString stringWithFormat: @"⏳ The rewarded \"%@\" is loading.", self.adUnitName];
}

- (void)showRewarded {
    if (!self.rewarded.isLoaded) {
        // Load rewarded one more time or notify a user about error
        [self loadRewarded];
        return;
    }
    
    [self.rewarded showFromViewController:self];
}

#pragma mark - PWFullScreenAdDelegate -

- (void)fullScreenAdDidLoad:(PWFullScreenAd * _Nonnull)ad {
    self.statusLabel.text = [NSString stringWithFormat: @"✅ The rewarded \"%@\" is loaded.", self.adUnitName];
    [self.getRewardButton setEnabled:YES];
}

- (void)fullScreenAdDidFailToLoad:(PWFullScreenAd * _Nonnull)ad {
    self.statusLabel.text = [NSString stringWithFormat: @"❌ Failed to load the rewarded \"%@\".", self.adUnitName];
}


- (void)fullScreenAdDidFailToPresentFullScreenContent:(PWFullScreenAd * _Nonnull)ad {
    self.statusLabel.text = [NSString stringWithFormat: @"❌ Failed to show the rewarded \"%@\".", self.adUnitName];
}

- (void)fullScreenAdWillPresentFullScreenContent:(PWFullScreenAd * _Nonnull)ad {}

- (void)fullScreenAdWillDismissFullScreenContent:(PWFullScreenAd * _Nonnull)ad {}

- (void)fullScreenAdDidDismissFullScreenContent:(PWFullScreenAd * _Nonnull)ad {
    self.rewarded = NULL;
    [self.getRewardButton setEnabled:NO];
}

- (void)fullScreenAdDidRecordClick:(PWFullScreenAd * _Nonnull)ad {}

- (void)fullScreenAdDidRecordImpression:(PWFullScreenAd * _Nonnull)ad {
    self.statusLabel.text = [NSString stringWithFormat: @"👍 The rewarded \"%@\" was successfully shown.", self.adUnitName];
}

- (void)fullScreenAdDidUserEarn:(PWFullScreenAd * _Nonnull)ad type:(NSString * _Nonnull)type amount:(double)amount {
    // Handle a reward regarding your business objectives.
    self.statusLabel.text = [NSString stringWithFormat: @"🎉 The reward is earned.\n Type: %@\n Amount: %0.f.", type, amount];
}

@end

File: PlaywireSDKAppsSwift/ads/fullscreen/rewarded/RewardedViewController.swift
//
//  RewardedViewController.swift
//  PlaywireDemo
//
//  Created by Playwire Mobile Team on 12/20/22.
//  Copyright © 2022 Playwire. All rights reserved.
//

import Playwire
import UIKit

final class RewardedViewController: UIViewController, AdUnitViewControllerType {
    
    @IBOutlet weak var statusLabel: UILabel!
    @IBOutlet weak var getRewardButton: UIButton!
    
    // The ad unit name, e.g. 'banner-320x50', 'interstitial-home', 'rewarded-coins', etc.
    var adUnitName: String!
    
    private var rewarded: PWRewarded?
    private var name: String { adUnitName ?? "NULL"}
    
    override func viewDidLoad() {
        super.viewDidLoad()
                
        getRewardButton.isEnabled = false
        loadRewarded()
    }
    @IBAction func getRewardAction(_ sender: UIButton) {
        showRewarded()
    }
    
    private func loadRewarded() {
        rewarded = PWRewarded(adUnitName: adUnitName, viewController: self, delegate: self)
        
        // Use `PWLoadParams().withTargeting()` to pass your custom targets to ad request.
        // let params = PWLoadParams().withTargeting(
        //  [
        //    "age": "18-32",
        //    "page": "travel"
        //  ]
        // )
        // rewarded?.load(params: params)
        
        rewarded?.load()
        statusLabel.text = "⏳ The rewarded \"\(name)\" is loading."
    }
    
    private func showRewarded() {
        guard rewarded?.isLoaded == true else {
            // Load rewarded one more time or notify a user about error
            loadRewarded()
            return
        }
        
        rewarded?.show()
    }
}

// MARK: - PWFullScreenAdDelegate -
extension RewardedViewController: PWFullScreenAdDelegate {
    func fullScreenAdDidLoad(_ ad: PWFullScreenAd) {
        statusLabel.text = "✅ The rewarded \"\(name)\" is loaded."
        getRewardButton.isEnabled = true
    }
    
    func fullScreenAdDidFailToLoad(_ ad: PWFullScreenAd) {
        statusLabel.text = "❌ Failed to load the rewarded \"\(name)\"."
    }
    
    func fullScreenAdWillPresentFullScreenContent(_ ad: PWFullScreenAd) {
        
    }
    
    func fullScreenAdWillDismissFullScreenContent(_ ad: PWFullScreenAd) {

    }
    
    func fullScreenAdDidDismissFullScreenContent(_ ad: PWFullScreenAd) {
        rewarded = nil
        getRewardButton.isEnabled = false
    }
    
    func fullScreenAdDidFailToPresentFullScreenContent(_ ad: PWFullScreenAd) {
        statusLabel.text = "❌ Failed to show the rewarded \"\(name)\"."
    }
    
    func fullScreenAdDidRecordImpression(_ ad: PWFullScreenAd) {
        statusLabel.text = "👍 The rewarded \"\(name))\" was successfully shown."
    }
    
    func fullScreenAdDidRecordClick(_ ad: PWFullScreenAd) {
        
    }
    
    func fullScreenAdDidUserEarn(_ ad: PWFullScreenAd, type: String, amount: Double) {
        // Handle a reward regarding your business objectives.
        statusLabel.text =  "🎉 The reward is earned.\n Type: \(type)\n Amount: \(String(format: "%.2f", amount))."
    }
}
